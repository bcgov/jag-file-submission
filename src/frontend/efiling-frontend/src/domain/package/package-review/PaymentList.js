/* eslint-disable react/jsx-one-expression-per-line, no-nested-ternary */
import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import Dinero from "dinero.js";
import { MdPrint } from "react-icons/md";
import "./PaymentList.scss";

export default function PaymentList({ payments }) {
  const dineroInit = {
    stat: Dinero({ amount: 0 }),
    toDate: Dinero({ amount: 0 }),
  };

  const [subtotal, setSubtotal] = useState(dineroInit);
  const [csoTotal, setCsoTotal] = useState(dineroInit);
  const [total, setTotal] = useState(dineroInit);
  const format = "$0,0.00";

  useEffect(() => {
    let subStat = Dinero({ amount: 0 });
    let subToDate = Dinero({ amount: 0 });
    let csoStat = Dinero({ amount: 0 });
    let csoToDate = Dinero({ amount: 0 });

    if (payments) {
      payments.forEach((payment) => {
        if (payment.paymentCategory === 1 || payment.paymentCategory === 2) {
          csoStat = csoStat.add(
            Dinero({ amount: payment.submittedAmount * 100 })
          );
          csoToDate = csoToDate.add(
            Dinero({ amount: payment.processedAmount * 100 })
          );
        } else if (!payment.feeExempt) {
          subStat = subStat.add(
            Dinero({ amount: payment.submittedAmount * 100 })
          );
          subToDate = subToDate.add(
            Dinero({ amount: payment.processedAmount * 100 || 0 })
          );
        }
      });
    }
    setSubtotal({ stat: subStat, toDate: subToDate });
    setCsoTotal({ stat: csoStat, toDate: csoToDate });
    setTotal({ stat: subStat.add(csoStat), toDate: subToDate.add(csoToDate) });
  }, [payments]);

  return (
    <div className="ct-payment-list">
      <table className="payment-table table table-borderless">
        <thead>
          <tr>
            <th scope="col">Document</th>
            <th scope="col" className="text-right">
              Statutory Fee
            </th>
            <th scope="col" className="text-right">
              Fees Charged to Date
            </th>
          </tr>
        </thead>
        <tbody>
          {payments &&
            payments
              .filter(
                (payment) =>
                  payment.paymentCategory !== 1 && payment.paymentCategory !== 2
              )
              .map((payment) => (
                <tr key={payment}>
                  <td>{payment.paymentDescription}</td>
                  <td className=" text-right">
                    {payment.feeExempt
                      ? Dinero({ amount: 0 }).toFormat(format)
                      : Dinero({
                          amount: payment.submittedAmount * 100,
                        }).toFormat(format)}
                  </td>
                  <td className="text-right">
                    {payment.feeExempt
                      ? "Exempt"
                      : payment.processedAmount
                      ? Dinero({
                          amount: payment.processedAmount * 100,
                        }).toFormat(format)
                      : "Pending"}
                  </td>
                  <td />
                </tr>
              ))}
          {payments && (
            <>
              <tr>
                <td className="text-right">Subtotal: </td>
                <td className="sub-table top-border">
                  {subtotal.stat.toFormat(format)}
                </td>
                <td className="sub-table top-border">
                  {subtotal.toDate.toFormat(format)}
                </td>
                <td />
              </tr>
              <tr>
                <td className="text-right">CSO Fees: </td>
                <td className="sub-table">{csoTotal.stat.toFormat(format)}</td>
                <td className="sub-table">
                  {csoTotal.toDate.toFormat(format)}
                </td>
                <td />
              </tr>
              <tr>
                <td className="text-right">Total Fees: </td>
                <td className="sub-table">{total.stat.toFormat(format)}</td>
                <td className="sub-table">{total.toDate.toFormat(format)}</td>
                <td />
              </tr>
            </>
          )}
        </tbody>
      </table>
      <div className="d-flex justify-content-end">
        <span className="file-href" role="button">
          View receipt
        </span>
        <MdPrint size="24" color="#7F7F7F" className="align-icon" />
      </div>
    </div>
  );
}

PaymentList.propTypes = {
  payments: PropTypes.arrayOf(PropTypes.object).isRequired,
};
