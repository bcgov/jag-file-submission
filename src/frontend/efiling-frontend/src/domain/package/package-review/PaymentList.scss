.ct-payment-list {
  td {
    padding: 7px;
  }
  .sub-table {
    font-weight: bold;
    text-align: right;
  }
  .top-border {
    border-top:  1px solid #989898;
  }
}
