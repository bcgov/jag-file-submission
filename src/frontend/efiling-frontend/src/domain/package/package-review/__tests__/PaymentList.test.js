import React from "react";
import { render, waitFor } from "@testing-library/react";
import PaymentList from "../PaymentList";

describe("PartyList Component", () => {
  beforeEach(() => {});

  test("Matches snapshot", () => {
    const payments = [
      {
        feeExempt: false,
        paymentCategory: 1,
        processedAmount: 7,
        submittedAmount: 7,
        serviceIdentifier: 0,
        transactionDate: null
      },
      {
        feeExempt: true,
        paymentCategory: 3,
        processedAmount: 7,
        submittedAmount: null,
        serviceIdentifier: 0,
        transactionDate: null
      },
      {
        feeExempt: false,
        paymentCategory: 3,
        processedAmount: 7,
        submittedAmount: 7,
        serviceIdentifier: 0,
        transactionDate: null
      },
      {
        feeExempt: false,
        paymentCategory: 3,
        processedAmount: null,
        submittedAmount: 7,
        serviceIdentifier: 0,
        transactionDate: null
      }
    ];
    const { asFragment } = render(<PaymentList payments={payments} />);
    waitFor(() => {});

    expect(asFragment()).toMatchSnapshot();
  });
});