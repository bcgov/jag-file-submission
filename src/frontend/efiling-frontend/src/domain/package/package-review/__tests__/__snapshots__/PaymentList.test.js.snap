// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartyList Component Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ct-payment-list"
  >
    <table
      class="payment-table table table-borderless"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Document
          </th>
          <th
            class="text-right"
            scope="col"
          >
            Statutory Fee
          </th>
          <th
            class="text-right"
            scope="col"
          >
            Fees Charged to Date
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td />
          <td
            class=" text-right"
          >
            $0.00
          </td>
          <td
            class="text-right"
          >
            Exempt
          </td>
          <td />
        </tr>
        <tr>
          <td />
          <td
            class=" text-right"
          >
            $7.00
          </td>
          <td
            class="text-right"
          >
            $7.00
          </td>
          <td />
        </tr>
        <tr>
          <td />
          <td
            class=" text-right"
          >
            $7.00
          </td>
          <td
            class="text-right"
          >
            Pending
          </td>
          <td />
        </tr>
        <tr>
          <td
            class="text-right"
          >
            Subtotal: 
          </td>
          <td
            class="sub-table top-border"
          >
            $14.00
          </td>
          <td
            class="sub-table top-border"
          >
            $7.00
          </td>
          <td />
        </tr>
        <tr>
          <td
            class="text-right"
          >
            CSO Fees: 
          </td>
          <td
            class="sub-table"
          >
            $7.00
          </td>
          <td
            class="sub-table"
          >
            $7.00
          </td>
          <td />
        </tr>
        <tr>
          <td
            class="text-right"
          >
            Total Fees: 
          </td>
          <td
            class="sub-table"
          >
            $21.00
          </td>
          <td
            class="sub-table"
          >
            $14.00
          </td>
          <td />
        </tr>
      </tbody>
    </table>
    <div
      class="d-flex justify-content-end"
    >
      <span
        class="file-href"
        role="button"
      >
        View receipt
      </span>
      <svg
        class="align-icon"
        color="#7F7F7F"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(127, 127, 127);"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 8H5c-1.66 0-3 1.34-3 3v6h4v4h12v-4h4v-6c0-1.66-1.34-3-3-3zm-3 11H8v-5h8v5zm3-7c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm-1-9H6v4h12V3z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
